### PR TITLE
implementation the ArrayList

### PR DESCRIPTION
**Pull request by BusyhinD**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,129 @@
...
-        return false;
+        return size == 0;
+    }
+
+    private static class Storage<T> {
+        private T object;
+
+        public Storage(T object) {
+            this.object = object;
+        }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:127]:
```suggestion
```
What is the reason for this class? Use simple T

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private Storage<T>[] dataStorage;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:8]:
```suggestion
    private T[] dataStorage;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,129 @@
...
+    private void indexCheck(int index) {
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index less than zero");
+        } else if (index != 0 && index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Not found element with index: " + index);
+        }
+    }
+
+    private void sizeCheck() {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:94]: sizeCheck is not correct name. You not only check size, you call grow() if needed. It must be clear from the method name.

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,129 @@
...
 
     @Override
     public void set(T value, int index) {
+        indexCheck(index);
+        dataStorage[index].object = value;
+    }
 
+    @Override
+    public T remove(T element) {
+        if (element == null) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:59]: You do not need outer IF - check null inside loop in IF statement.

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,129 @@
...
+        size--;
+        System.arraycopy(dataStorage, index + 1, dataStorage, index, size - index);
+        dataStorage[size] = null;
+        return object;
     }
 
-    @Override
-    public T remove(T element) {
-        return null;
+    private void indexCheck(int index) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:85]: One IF is enough for this method. And create informative message with INDEX and SIZE

Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:85]:
```suggestion
    private void checkIndex(int index) {
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,116 @@
...
-
+        if (list.size() > dataStorage.length - size) {
+            grow(size + list.size());
+        }
+        T[] newStorage = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            newStorage[i] = list.get(i);
+        }
+        System.arraycopy(newStorage, 0, dataStorage, size, list.size());
+        size += list.size();
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:42]:
```suggestion
        for (int i = 0; i < list.size(); i++) {
            dataStorage[size++] = list.get(i);
        }
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,116 @@
...
+        int oldCapacity = dataStorage.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        grow(newCapacity);
+    }
+
+    private void grow(int newCapacity) {
+        T[] newStorage = (T[]) new Object[newCapacity];
+        System.arraycopy(dataStorage, 0, newStorage, 0, size);
+        dataStorage = newStorage;
     }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:106]: All public methods should be before private ones in your class.

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,113 @@
...
+        for (int i = 0; i < size; i++) {
+            if (dataStorage[i] == element) {
+                return remove(i);
+            }
+        }
+        for (int i = 0; i < size; i++) {
+            if (element != null && element.equals(dataStorage[i])) {
+                return remove(i);
+            }
+        }
```
Comment (kateryna-deshchinska) [inline src/main/java/core/basesyntax/ArrayList.java:65]: compare this two cycle to one

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,113 @@
...
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list.size() > dataStorage.length - size) {
+            grow(size + list.size());
+        }
+        for (int i = 0; i < list.size(); i++) {
+            dataStorage[size++] = list.get(i);
+        }
```
Comment (kateryna-deshchinska) [inline src/main/java/core/basesyntax/ArrayList.java:39]:
```suggestion
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```
you can do like this

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,113 @@
...
+        if (size == dataStorage.length) {
+            grow();
+        }
+    }
+
+    private void grow() {
+        int oldCapacity = dataStorage.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        grow(newCapacity);
+    }
```
Comment (kateryna-deshchinska) [inline src/main/java/core/basesyntax/ArrayList.java:106]: You can merge this two methods to one)

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,113 @@
...
+
+    private void grow(int newCapacity) {
+        T[] newStorage = (T[]) new Object[newCapacity];
+        System.arraycopy(dataStorage, 0, newStorage, 0, size);
+        dataStorage = newStorage;
     }
```
Comment (kateryna-deshchinska) [inline src/main/java/core/basesyntax/ArrayList.java:112]: you can move this logic inside grow() method

**Review by kateryna-deshchinska — CHANGES_REQUESTED** [review]

**Review by aaananas — APPROVED** [review]

**Pull request closed without merging**